test/host_tests/sal_tcpclient.py

```diff
@@ -0,0 +1,110 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import sys
+import socket
+import time
+import threading
+import re
+from mbed_host_tests import BaseHostTest
+
+
+class SalTcpClientTest(BaseHostTest):
+    """
+    mbed greentea framework host test script for tcp_echo_server client 
+    side functionality. The test does the following:
+      - Receives over the serial port the target ipaddr, port of the 
+        target tcp server
+      - establishes a connection with the remoted tcp echo server
+      - sends ever increasing data buffers using send() starting with
+        a 16 byte buffer and doubling the size of the data sent on each 
+        iteration through the loop. 
+      - When finished, a command is send to the remote tcp server to 
+        cause it to terminate. 
+    """
+
+    name = 'sal_tcpclient'
+
+    def tcp_client_tx_rx(self, ipaddrs, port):
+        "Thread entrypoint for sending and receiving data to/from the remote tcp echo server."
+      
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.bind(("", 0))
+        s.connect((ipaddrs, port))
+        
+        tx_data = "1234567890abcdef"
+        for i in xrange(8):
+            s.send(tx_data)
+            print "Host: sent %u bytes of data" % len(tx_data)
+            rx_data = s.recv(len(tx_data))
+            tx_data += tx_data
+            print "Host: recevied %u bytes of data" % len(rx_data)
+
+        # command to get the remove tcp server to terminate
+        data = "quit"
+        s.send(data)
+        # let remote target terminate the tcp connection
+        while True:
+            data = s.recv(1024)
+            if not data: 
+                break
+        s.close()
+
+        
+    def test(self, selftest):
+        # Returning none will suppress host test from printing success code
+        """ main greentea host test script entry point which implements the functionality of the test as described above."""
+        
+        # mbed greentea framwork support the following for outputing serial console emitted 
+        # from the target, but it doesnt work relibly for me.
+        # selftest.dump_serial()
+        # selftest.dump_serial_end()
+        
+        # until its fixed, the following emits the serial console trace
+        while True:
+            c = selftest.mbed.serial_readline()
+            if c is None:
+                selftest.print_result(self.RESULT_IO_SERIAL)
+                return
+            
+            print "MBED: " + c.strip()
+            
+            # look for the end tag in serial output denoting the test has 
+            # finished, and this can return.
+            if c.strip() == "{{end}}":
+                print "HOST: Terminating Test"
+                break
+            
+            # find remote tcp server IP address in serial console stream:
+            elif ">>> EC" in c.strip():
```

Comment:
it could be updated to utest, but the requirement was to get the tests working. There was insufficient time on the ticket to port to utest too.